### PR TITLE
Link between Genesis and Kraftwerk

### PR DESCRIPTION
**Pull request by loichenninger**

### CODE REVIEW

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/sequence/BuildBindingsSequenceGenesis.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+			for (ExternalVariable extVar : surveyUnit.getExternalVariables()){
+				// The external are always in root group name
+				if (!extVar.getValues().isEmpty()) answers.putValue(extVar.getIdVar(), extVar.getValues().get(0));
+			}
+
+			data.getQuestionnaires().add(questionnaire);
+		}
+
+//		/* Step 2.2 : Get paradata for the survey */
```
Comment (alexisszmundy) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/sequence/BuildBindingsSequenceGenesis.java:58]: Indiquer pourquoi c'est commenté ou supprimer (Sonar aime pas trop les blocs de code commenté)

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/sequence/BuildBindingsSequenceGenesis.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+//
+//		/* Step 2.3 : Get reportingData for the survey */
+//		parseReportingData(modeInputs, data);
+
+		/* Step 2.4a : Convert data object to a VTL Dataset */
+		data.setDataMode(dataMode);
+		vtlExecute.convertToVtlDataset(data, dataMode, vtlBindings);
+	}
+
+//	private void parseParadata(ModeInputs modeInputs, SurveyRawData data) throws NullException {
```
Comment (alexisszmundy) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/sequence/BuildBindingsSequenceGenesis.java:69]: Idem

**Review by alexisszmundy — COMMENTED** [review]
Petit code smell mais OK sinon

**Comment by alexisszmundy** [comment]
La PR affecte la branche main, on envoie pas sur la develop avant ?

File: kraftwerk-api/src/main/java/fr/insee/kraftwerk/api/client/GenesisClient.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		return modes;
+	}
+
+	public List<SurveyUnitUpdateLatest> getUELatestState(String idQuestionnaire, SurveyUnitId suId) {
+		String url = String.format("%s/response/get-simplified-response/by-ue-and-questionnaire/latest?idQuestionnaire=%s&idUE=%s", configProperties.getGenesisUrl(), idQuestionnaire, suId.getIdUE());
+		ResponseEntity<SurveyUnitUpdateLatest[]> response = restTemplate.exchange(url, HttpMethod.GET, null, SurveyUnitUpdateLatest[].class);
+		return response.getBody() != null ? Arrays.asList(response.getBody()) : null;
+	}
+
+	public List<SurveyUnitUpdateLatest> getUEsLatestState(String idQuestionnaire, List<SurveyUnitId> idUEs) {
```
Comment (alicela) [inline kraftwerk-api/src/main/java/fr/insee/kraftwerk/api/client/GenesisClient.java:53]: change name to be coherent with "POST"

File: kraftwerk-api/src/main/java/fr/insee/kraftwerk/api/services/MainService.java [hunk 4]
```diff
@@ -84,6 +95,18 @@ public ResponseEntity<String> mainLunaticOnly(
 		return ResponseEntity.ok(inDirectoryParam);
 	}
 
+	@PutMapping(value = "/main/genesis")
+	@Operation(operationId = "mainGenesis", summary = "${summary.mainGenesis}", description = "${description.mainGenesis}")
+	public ResponseEntity<String> mainGenesis(@RequestParam("idQuestionnaire") String idQuestionnaire) throws KraftwerkException, IOException {
```
Comment (alicela) [inline kraftwerk-api/src/main/java/fr/insee/kraftwerk/api/services/MainService.java:100]: Kraftwerk exception is catched

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/data/model/Mode.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+	private final String folder;
+
+	Mode(String modeName, String folder) {
+		this.modeName = modeName;
+		this.folder = folder;
+	}
+
+	public static Mode getEnumFromModeName(String modeName) {
+		for (Mode mode : Mode.values()) {
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/data/model/Mode.java:20]: https://stackoverflow.com/questions/27807232/finding-enum-value-with-java-8-stream-api

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/data/model/SurveyUnitUpdateLatest.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+public class SurveyUnitUpdateLatest {
+
+	private String idQuest;
+	private String idCampaign;
+	private String idUE;
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/data/model/SurveyUnitUpdateLatest.java:12]: String idUE or SurveyUnitId ?

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/Mode.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+	private final String folder;
+
+	Mode(String modeName, String folder) {
+		this.modeName = modeName;
+		this.folder = folder;
+	}
+
+	public static Mode getEnumFromModeName(String modeName) {
+		for (Mode mode : Mode.values()) {
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/Mode.java:20]: same comment : https://stackoverflow.com/questions/27807232/finding-enum-value-with-java-8-stream-api

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+public class UserInputsFile extends UserInputs {
+
+	@Getter
+	@Setter
+	private Path userInputFile;
+
+	private final Set<String> mandatoryFields = Set.of("survey_data", "data_mode", "data_file", 			"data_format", "multimode_dataset_name");
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java:32]: white spaces

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+			throw e;
+		}
+	}
+
+	public List<String> getModes() {
+		return new ArrayList<>(modeInputsMap.keySet());
+	}
+
+	private String readField(JsonNode node, String field) throws MissingMandatoryFieldException {
+		JsonNode value = node.get(field);
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java:88]: Refactor this method to reduce if else

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	private Path convertToPath(String userField) throws KraftwerkException {
+		if (userField != null && !"null".equals(userField) && !"".equals(userField)) {
+			Path inputPath = inputDirectory.resolve(userField);
+			if (!new File(inputPath.toUri()).exists()) {
+				throw new KraftwerkException(400, String.format("The input folder \"%s\" does not exist in \"%s\".", userField, inputDirectory.toString()));
+			}
+			return inputPath;
+		} else {
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java:119]: Else is not necessary here

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+			return null;
+		}
+	}
+
+	private Path convertToUserPath(String userField) {
+		if (userField != null && !"null".equals(userField) && !"".equals(userField)) {
+			return Paths.get(userField);
+		} else {
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java:127]: same comment

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+	private URL convertToUrl(String userField) {
+		if (userField == null) {
+			log.debug("null value out of method that reads DDI field (should not happen).");
+			return null;
+		}
+		try {
+			if (userField.startsWith("http")) {
+				return new URL(userField);
+			} else {
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsFile.java:140]: same comment

File: kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsGenesis.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+		if (value != null) {
+			String text = value.asText();
+			if (!(text.equals("") || text.equals("null"))) {
+				return text;
+			}
+		}
+		return null;
+	}
+
+	private Path convertToPath(String userField) throws KraftwerkException {
```
Comment (alicela) [inline kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsGenesis.java:112]: convertToPath, convertToUrl in the other class = maybe a class utils

Comment (loichenninger) [reply kraftwerk-core/src/main/java/fr/insee/kraftwerk/core/inputs/UserInputsGenesis.java]: I moved the two methods in FileUtils

**Pull request merged**